### ISSUE TITLE
Fetch-mock in the browser requires a global fetch implementation

### ISSUE THREAD

**Issue by pimterry** [wheresrhys/fetch-mock#158]
In the browser code `fetch-mock` assumes that Request, Response and Header globals are available, which means non-global development with fetch isn't possible with fetch-mock in browsers.

Our current workaround is to just expose all these globals in the tests:

```javascript
if (IS_BROWSER) {
	realFetchModule = require('fetch-ponyfill')({ Promise: BluebirdPromise })
	global.Headers = realFetchModule.Headers
	global.Request = realFetchModule.Request
	global.Response = realFetchModule.Response
}
```

It'd be nice to not have to do that, since it's a little messy, and because this clearly isn't necessary, as the Node implementation doesn't make these assumptions and works fine without these globals.

This would be a fairly niche problem, except that PhantomJS doesn't support fetch, and that's where most browser automated testing happens.

**Comment by wheresrhys** [comment]
My issue with this is that `fetch` and its associated classes are supposed to be global in the browser https://fetch.spec.whatwg.org/, so its a pretty reasonable assumption that the user's app takes care of making these globals available before trying to run either tests or application code. However one thing I'm increasingly aware of is that the use cases for fetch-mock are quite varied - and ponyfills are a lot more in vogue now than a few months ago - so I'm open to making your use case easier, but with the proviso that the default behaviour stays as it is.

Do you have any suggestions? Is there any reason you can't use one of these approaches https://github.com/wheresrhys/fetch-mock#polyfilling-fetch

**Comment by pimterry** [comment]
The main problem for us is that we're trying to provide modules that will be used in other people's code. We don't want to write new globals for them unexpectedly (e.g. by including isomorphic-fetch automatically), but we'd also like to make it really easy to pick up and use these libraries (so I'd like to avoid 'Install our library, and also remember to install and require isomorphic-fetch yourself in advance if you're in Node or an old browser' instructions everywhere).

We're building a library, but we're not developing the end user application, so we can't dictate what exists in the resulting environment.

We could polyfill the environment more actively for our tests (and that's basically what we're doing in the workaround above), it's just that it'd be nice to test an environment that's more representative of our end user case. They've only imported the library, and they may or may not have fetch available, and our library should Just Work. This isn't desperately urgent or anything, you can work around it, but it will be a stumbling block for anybody else doing this too.

In terms of solutions, I'd suggest tying this into sandbox again to keep the existing behaviour normally, so you can write:

```javascript
// This is { fetch: ..., Headers: ..., Request: ..., Response: ... }
var fetchPonyfill = require('fetch-ponyfill')({ Promise: Promise })

// Pass an object that contains all these globals in, and pull out the ones you need
var fetch = require('fetch-mock').sandbox(Promise, fetchPonyfill)
```

**Comment by wheresrhys** [comment]
Yep that all seems reasonable. It's a really weird function signature, but if the common use case is using fetch-ponyfill then your solution would be easier for the user. Can always add duck-typed single options object later.

D'you want to submit another PR. I'm also thinking that the README could do with significant additions to cover the polyfill/ponyfill differences. In fact, it's probably time I added a `/docs` directory, but in the meantime please add subsections to the Quickstart if you want to.

**Comment by wheresrhys** [comment]
Actually hold fire on this for a bit. Is there a case for making this work when not using `.sandbox()`? If so it might make more sense to expose the FetchMock.setGlobals static method to the end user, as this already has provision for setting globals https://github.com/wheresrhys/fetch-mock/blob/master/src/client.js#L7. Potentially worth adding Promise customisation in there too. Don't want to muddy the waters between telling fetch-mock what is available as a global and telling fetch-mock what it should use though, so will think about it some more first. Possibly just renaming the method to `setThings`*

*todo - find better term than things

**Comment by wheresrhys** [comment]
`setImplementations` is gonna be my slightly clunky sounding but quite descriptive method name

**Comment by wheresrhys** [comment]
released in 5.9.0
